src/stream/src/executor/top_n/top_n_cache.rs

```diff
@@ -861,3 +807,86 @@ impl AppendOnlyTopNCacheTrait for TopNCache<true> {
         Ok(())
     }
 }
+
+/// Used to manage staging changes for one `TopNCache` (for one group).
+/// It should be maintained when an entry is inserted or deleted from the `middle` cache.
+#[derive(Debug, Default)]
+pub struct TopNStaging {
+    to_delete: BTreeMap<CacheKey, CompactedRow>,
+    to_insert: BTreeMap<CacheKey, CompactedRow>,
+    to_update: BTreeMap<CacheKey, (CompactedRow, CompactedRow)>,
+}
+
+impl TopNStaging {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    fn insert(&mut self, cache_key: CacheKey, row: CompactedRow) {
+        if let Some(old_row) = self.to_delete.remove(&cache_key) {
+            if old_row != row {
+                self.to_update.insert(cache_key, (old_row, row));
+            }
+        } else {
+            debug_assert!(!self.to_update.contains_key(&cache_key));
+            debug_assert!(!self.to_insert.contains_key(&cache_key));
+            self.to_insert.insert(cache_key, row);
```

Comment:
Shall we enable assertion in production as well, just like mem-table conflict?

Reply:
Actually I realized that we shouldn't do any assertion in this structure. We should leave the responsibility to `TopNCache`, and the latter, instead of simply `assert`, will do `consistency_error`/`consistency_panic`.